bluemira/equilibria/fem_fixed_boundary/fem_magnetostatic_2D.py

```diff
@@ -183,3 +196,234 @@ def calculate_b(self):
         self.B = dolfin.project(dolfin.as_vector((Bx, Bz)), w)
 
         return self.B
+
+
+class FemGradShafranovFixedBoundary(FemMagnetostatic2d):
+    """
+    A 2D fem Grad Shafranov solver. The solver is thought as support for the fem fixed
+    boundary module.
+
+    Parameters
+    ----------
+    mesh : dolfin.mesh or string
+           the filename of the xml file with the mesh definition
+           or a dolfin mesh
+    boundaries : dolfin.MeshFunction or string
+                 the filename of the xml file with the boundaries definition
+                 or a MeshFunction that defines the boundaries
+    p_order : int
+        the order of the approximating polynomial basis functions
+    """  # noqa (W505)
+
+    def __init__(self, mesh, boundaries=None, p_order=3):
+        super().__init__(mesh, boundaries, p_order)
+        self.k = 1
+
+    @property
+    def psi_ax(self):
+        """Poloidal flux on the magnetic axis"""
+        return np.max(self.psi.vector()[:])
+
+    @property
+    def psi_b(self):
+        """Poloidal flux on the boundary"""
+        return np.min(self.psi.vector()[:])
+
+    @property
+    def psi_norm_2d(self):
+        """Normalized flux function in 2-D"""
+
+        def myfunc(x):
+            value = np.sqrt(
+                np.abs((self.psi(x) - self.psi_ax) / (self.psi_b - self.psi_ax))
+            )
+            return value
+
+        return myfunc
+
+    def _create_g_func(
+        self,
+        pprime: Union[Callable, float],
+        ffprime: Union[Callable, float],
+        curr_target: float,
+    ):
+        """
+        Returns the density current function given pprime and ffprime.
+
+        Parameters
+        ----------
+        pprime: Union[Callable, float]
+            pprime as function of psi_norm (1-D function)
+        ffprime: Union[Callable, float]
+            ffprime as function of psi_norm (1-D function)
+        curr_target: float
+            target current (also used to initialize the solution in case self.psi is
+            still 0 and pprime and ffprime are, then, not defined)
+
+        Returns
+        -------
+        g: Callable
+            source current to solve the magnetostatic problem
+        """
+        dx = dolfin.Measure("dx", domain=self.mesh)
+        area = dolfin.assemble(dolfin.Constant(1) * dx())
```

Comment:
I'm not sure if this is the case, but it looks like this is where the first initial guess of the current distribution is made (i.e. constant current across the cells (??)). If this is the case, this is probably where the lowest hanging fruit is: a good starting guess makes a big difference for convergence.

Here's a potential pseudo-code way of getting a good starting guess:

* estimate the current centroid (either via a 0-D Shafranov shift estimate, or by some other means - I have seen a couple of approaches that could be useful here)
* use the flux functions and some estimate of psi_norm to calculate an initial current distribution

I've probably not fully wrapped my head around this yet though. From looking at the behaviour of the G-S iterations, though, it looks like the starting guess is not so good.

Reply:
OK, so I've played around with this a little bit by emulating a current distribution with strong central peaking:

```python
        grid_x, grid_z = np.array(self.mesh.coordinates()).T
        min_x = np.min(grid_x)
        max_x = np.max(grid_x)
        min_z = np.min(grid_z)
        max_z = np.max(grid_z)
        r_major = min_x + 0.5 * (max_x - min_x)+0.5
        z_major = 0.0
        max_distance = np.hypot(r_major-min_x, z_major-max(abs(min_z), max_z))

        def g_initial(x):
            r = x[0]
            z = x[1]
            distance = np.hypot((r_major-r), (z_major-z))
            return max(0, 2 - np.exp(distance / max_distance))

        def g(x):
            if self.psi_ax == 0:
                return g_initial(x)
            else:
                r = x[0]
                ...
```

I also tried `(return 1 - np.sqrt(distance/max_distance)`

Two things to note:
* With the function above on the first iteration in the example the psi on axis is much closer to the converged psi on axis (172.2 V.s on first iteration vs. 159.05 V.s using the constant current density)
* This has almost no effect on the number of iterations required to reach convergence.

I am quite surprised by this! One could play around with some better dolfin functions and try and get the integral of the first guess right (if it isn't somehow adjust in Fenics already), but it seems to me like this low-hanging fruit brings nothing!

Reply:
Potential future work, but not necessary at present